solana-programs/payment-router/programs/payment-router/src/utils.rs

```diff
@@ -0,0 +1,129 @@
+    // i.e. that the owner of the sender account is owned by the program.
+    // This is because we use the account macro with seeds and bump for the 'sender_owner'.
+    let sender_data = sender.data.borrow();
+    let sender_account_owner= <anchor_spl::token::spl_token::state::Account as anchor_spl::token::spl_token::state::GenericTokenAccount>
+        ::unpack_account_owner(&sender_data)
+        .unwrap();
+    if sender_account_owner != sender_owner.key {
+        return Err(PaymentRouterErrorCode::SenderTokenAccountNotOwnedByPDA.into());
+    }
+
+    // 2. Verify that the passed in token mint matches the sender token mint.
+    let token_mint = get_token_mint(is_audio);
+    let sender_mint = <anchor_spl::token::spl_token::state::Account as anchor_spl::token::spl_token::state::GenericTokenAccount>
+        ::unpack_account_mint(&sender_data)
+        .unwrap();
+    if sender_mint.key().to_string() != token_mint.to_string() {
+        return Err(PaymentRouterErrorCode::InvalidSenderTokenMint.into());
+    }
+
+    Ok(())
+}
+
+/**
+ * 1. Verify that the number of recipients matches the number of amounts.
+ * 2. Verify that the total of individual amounts matches the total amount.
+ */
+pub fn check_recipient_amounts(
+  remaining_accounts: &[AccountInfo],
+  amounts: Vec<u64>,
+  total_amount: u64
+) -> Result<()> {
+    // 1. Verify that the number of recipients matches the number of amounts.
+    if remaining_accounts.len() != amounts.len() {
+        return Err(PaymentRouterErrorCode::RecipientAmountMismatch.into());
+    }
+    // 2. Verify that the total of individual amounts matches the total amount.
+    let total: u64 = amounts.iter().sum();
+    if total != total_amount {
+        return Err(PaymentRouterErrorCode::TotalAmountMismatch.into());
+    }
+
+    Ok(())
+}
+
+/**
+ * Iterate through the recipient accounts, represented by the remaining accounts
+ * and transfer the respective amount to each recipient,
+ * while checking that the recipient account token mint matches the given token mint.
+ */
+pub fn execute_transfers<'info>(
+    sender: AccountInfo<'info>,
+    sender_owner: AccountInfo<'info>,
+    remaining_accounts: &[AccountInfo<'info>],
+    amounts: Vec<u64>,
+    is_audio: bool,
+    payment_router_pda_bump: u8,
+) -> Result<()> {
+    let remaining_accounts_iter = &mut remaining_accounts.iter();
+    let mut amount_index = 0;
+    let token_mint = get_token_mint(is_audio);
+    while let Ok(receiver) = next_account_info(remaining_accounts_iter) {
+        let receiver = receiver.to_account_info();
+
+        // Verify that the passed in token mint matches the sender token mint.
+        let receiver_data = receiver.data.borrow();
+        let receiver_mint = <anchor_spl::token::spl_token::state::Account as anchor_spl::token::spl_token::state::GenericTokenAccount>
+            ::unpack_account_mint(&receiver_data)
+            .unwrap();
```

Comment:
Panic

Reply:
tell me more. i think we're okay if this panics and fails